ros2cli/ros2cli/daemon/__init__.py

```diff
@@ -92,6 +96,65 @@ def shutdown_handler():
             server.server_close()
 
 
+class _DirectNode:
+    """A direct node, that resets itself when a network interface changes."""
+
+    def __init__(self, args):
+        self.args = args
+        # TODO(ivanpauno): A race condition is possible here, since it isn't possible to know
+        # exactly which interfaces were available at node creation.
+        self.node = DirectNode(args)
+        self.addresses_at_start = self.interfaces_ip_addresses()
+
+    def __enter__(self):
+        self.node.__enter__()
+        return self
+
+    def __getattr__(self, name):
+        attr = self.node.__getattr__(name)
+
+        def wrapper():
+            self.reset_if_addresses_changed()
+            return self.node.__getattr__(name)()
+        wrapper.__name__ = attr.__name__
+
+        if inspect.ismethod(attr):
+            return wrapper
```

Comment:
@ivanpauno consider:
```python
if inspect.ismethod(attr):
    @functools.wraps(attr)
    def wrapper(*args, **kwargs):
        self.reset_if_addresses_changed()
        return getattr(self.node, name)(*args, **kwargs)
    wrapper.__signature__ = inspect.signature(attr)
```
instead

Reply:
Is this needed?
`wrapper.__signature__ = inspect.signature(attr)`

Reply:
AFAIK, `functools.wraps` does that automatically.

Reply:
It is needed if we want to keep the decorator as transparent as possible, and no, `functools.wraps` does not force any function signature, only naming.

Reply:
See https://github.com/ros2/launch/pull/279#discussion_r335518599.

Reply:
See https://github.com/ros2/launch/pull/279#discussion_r335626803, though it's arguably a detail that has no impact on this particular code (at least for the time being).

Reply:
As a proof, you run do the following:
```python3
import functools
import inspect

def asd(bsd, csd):
    print('asd')

@functools.wraps(asd)
def wrapper(*args, **kwargs):
    asd(*args, **kwargs)

print(inspect.signature(wrapper) == inspect.signature(asd))
```

The result is `True`.